test/version_solver_test.dart

```diff
@@ -1300,6 +1301,241 @@ void downgrade() {
   });
 }
 
+void features() {
+  test("doesn't enable a feature by default", () async {
+    await servePackages((builder) {
+      builder.serve('foo', '1.0.0', pubspec: {
+        "features": {
+          "feature": {
+            "dependencies": {'bar': '1.0.0'}
+          }
+        }
+      });
+      builder.serve('bar', '1.0.0');
+    });
+
+    await d.appDir({'foo': '1.0.0'}).create();
+    await expectResolves(result: {'foo': '1.0.0'});
+  });
+
+  test("enables a feature if it's required", () async {
+    await servePackages((builder) {
+      builder.serve('foo', '1.0.0', pubspec: {
+        "features": {
+          "feature": {
```

Comment:
Nit: Can you use a slightly less confusing name for the feature? Maybe just "foof"?